forest/cli/src/cli/snapshot_cmd.rs

```diff
@@ -323,6 +348,124 @@ fn clean(config: &Config, snapshot_dir: &Option<PathBuf>, force: bool) -> anyhow
     Ok(())
 }
 
+async fn validate(
+    config: &Config,
+    validate_height: &i64,
+    snapshot: &PathBuf,
+) -> anyhow::Result<()> {
+    let confirm = Confirm::with_theme(&ColorfulTheme::default())
+        .with_prompt(format!(
+            "This will result in using approximately {} MB of data. Proceed?",
+            std::fs::metadata(snapshot).unwrap().len() / (1024 * 1024)
+        ))
+        .default(false)
+        .interact()
+        .unwrap_or_default();
+
+    if confirm {
+        // clear temp db if exists
+        let tmp_db_path = PathBuf::from("/tmp/validate_snapshot");
+        let _ = std::fs::remove_dir_all(&tmp_db_path);
+
+        let db_path = tmp_db_path.join(&config.chain.name);
+        // 1. load car into a temporary database
+        let db = forest_db::rocks::RocksDb::open(db_path, &config.rocks_db)
+            .expect("Opening RocksDB must succeed");
+
+        let chain_store = Arc::new(ChainStore::new(db).await);
+
+        let genesis = read_genesis_header(
+            config.client.genesis_file.as_ref(),
+            config.chain.genesis_bytes(),
+            &chain_store,
+        )
+        .await
+        .unwrap();
+        let genesis_header = &genesis.blocks()[0];
+        chain_store.set_genesis(genesis_header).unwrap();
+
+        let cids = {
+            let file = tokio::fs::File::open(&snapshot).await?;
+            let reader = FetchProgress::fetch_from_file(file).await?;
+            load_car(&chain_store.db, reader.compat()).await?
+        };
+
+        let ts = chain_store.tipset_from_keys(&TipsetKeys::new(cids)).await?;
+
+        let snapshot_head_epoch = ts.epoch();
+
+        chain_store.db.flush()?;
+
+        validate_links_and_genesis_traversal(&chain_store, ts, &chain_store.db, *validate_height)
+            .await
+            .expect(&format!(
+                "failed validating ipld links and genesis traversal from  snapshot at height: {}",
+                &snapshot_head_epoch
+            ));
+
+        // clear temp db path contents
+        let _ = std::fs::remove_dir_all(&tmp_db_path);
+    }
+
+    Ok(())
+}
+
+async fn validate_links_and_genesis_traversal<DB>(
+    chain_store: &ChainStore<DB>,
+    ts: Arc<Tipset>,
+    db: &DB,
+    upto: ChainEpoch,
+) -> anyhow::Result<()>
+where
+    DB: fvm_ipld_blockstore::Blockstore + Store + Send + Sync + Clone,
+{
+    let mut seen = std::collections::HashSet::<Cid>::new();
+    let upto = ts.epoch() - upto;
+
+    println!(
+        "validating cids from snapshot in db: {} <- {} ...",
+        upto,
+        ts.epoch()
+    );
+
+    let mut ts = ts.parents().clone();
+
+    loop {
+        let h = chain_store.tipset_from_keys(&ts).await?;
+        let height = h.epoch();
+
+        if height == 0 {
+            break;
```

Comment:
Let's verify that we reached the expected genesis tipset. Ie. validating a calibnet snapshot as if it was a mainnet snapshot should fail.